pypicloud/cache/dynamo.py

```diff
@@ -22,6 +25,78 @@
 LOG = logging.getLogger(__name__)
 
 
+def _path_depth(package):
+    """Count the slashes in path, or return infinity on no path."""
+    path = package.data.get('path')
+    if not path:
+        return float('+inf')
+    return path.count('/')
+
+
+def _decide_between_versions(contender, current):
+    """Decide between packages with the same filename on different paths."""
+
+    if contender.last_modified < current.last_modified:
+        return contender
```

Comment:
Wow, this is being particularly hard to keep all the details at once in my head. However, here's what I currently think is our best approach:

```
def _decide_between_versions(cache, storage):
  if cache.filename != storage.filename:
    return current  # Trivial decision; pretty sure we agree on this.

  if cache.data['path'] == storage.data['path']:
    return contender  # Storage has new data for us on this path. Just accept it.

  return sorted(
    [cache, storage],
    key=lambda o: (
      o.last_modified,
      o.data['path'],
    ),
  )[0]
```

I'm a bunch lost, and I might be leaving some case behind, but, I think this nails it all. The choice of using path (not using depth anymore) is pretty arbitrary. The only thing I need is a way to choose between different files that were uploaded at *exactly* the same second to S3. Since we've already asserted they're different, and strings are fully orderable, one of them has to always win the same way. The only way (I see) to change the file for a filename is to delete it. I can live with that.

Let me know if I stopped making sense somewhere or if I contradicted myself again. If we need to iterate this further, I vote we start writing down the properties we want observed as test cases. This way, we can have a machine tell us what we left behind.